### PR TITLE
[xcm] runtime api for LocationToAccount conversions

### PR DESCRIPTION
**Pull request by bkontur**
Closes: https://github.com/paritytech/polkadot-sdk/issues/4298

This PR also merges `xcm-fee-payment-runtime-api` module to the `xcm-runtime-api`.

## TODO

- [x] rename `convert`  to `convert_location` and add new one `convert_account` (opposite direction)
- [x] add to the all testnet runtimes
- [x] check polkadot-js if supports that automatically or if needs to be added manually https://github.com/polkadot-js/api/pull/5917
- [ ] backport/patch for fellows and release (asap)

## Open questions
- [x] should we merge `xcm-runtime-api`  and `xcm-fee-payment-runtime-api` to the one module `xcm-runtime-api` ?

## Usage
Input:
 - `location:  VersionedLocation`
 
Output:
 - account_id bytes

image

### CODE REVIEW

**Review by BabakArtist — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]
Looks good!

> should we merge xcm-runtime-api and xcm-fee-payment-runtime-api to the one module xcm-runtime-api ?

that sounds good hygiene to me, makes sense to have a structure like:
```
xcm-runtime-api
|    mod conversions
|    mod dry_run
|    mod fees
```

File: polkadot/xcm/xcm-runtime-api/src/conversions.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Contains runtime APIs for useful conversions, such as between XCM `Location` and `AccountId`.
```
Comment (bkchr) [inline polkadot/xcm/xcm-runtime-api/src/conversions.rs:17]: There already exists multiple runtime apis for XCM. We should merge them all into one crate?

File: polkadot/xcm/xcm-runtime-api/src/conversions.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	/// Converting a versioned data structure from one version to another failed.
+	#[codec(index = 1)]
+	VersionedConversionFailed,
+}
+
+sp_api::decl_runtime_apis! {
+	/// API for useful conversions between XCM `Location` and `AccountId`.
+	pub trait LocationToAccountApi {
+		/// Converts `Location` to `Account` with `AccountId` and Ss58 representation.
+		fn convert_location(location: VersionedLocation, ss58_prefix: Option<u16>) -> Result<Account, Error>;
```
[the remark this answers is not in the document]
Comment (bkchr) [reply polkadot/xcm/xcm-runtime-api/src/conversions.rs:59]: Creating an address out of the accountid is quite easy for the UI.

[the remark this answers is not in the document]
Comment (bkontur) [reply polkadot/xcm/xcm-runtime-api/src/conversions.rs:59]: > Creating an address out of the accountid is quite easy for the UI.

yes, there are several utilities for that:
https://polkadot.subscan.io/tools/format_transform
https://polkadot-address-convertor.netlify.app/

I just wanted to have it also here in one place, but I will remove SS58 stuff, no problem to simplify it here :)

File: polkadot/xcm/xcm-runtime-api/src/conversions.rs (hunk 1 above)
Comment (bkontur) [reply polkadot/xcm/xcm-runtime-api/src/conversions.rs:17]: yes, exactly, that is the open question from description, and also Adrian above is ok with that,
so I will merge them

Comment (bkontur) [reply polkadot/xcm/xcm-runtime-api/src/conversions.rs:17]: fixed here: https://github.com/paritytech/polkadot-sdk/pull/4857/commits/ab63f0dd2fc713c7ff45e4ecb58db076a223b600

File: cumulus/parachains/runtimes/collectives/collectives-westend/src/lib.rs [hunk 3]
```diff
@@ -163,6 +163,7 @@ parameter_types! {
 		})
 		.avg_block_initialization(AVERAGE_ON_INITIALIZE_RATIO)
 		.build_or_panic();
+	pub const SS58Prefix: u8 = 0;
```
Comment (bkchr) [inline cumulus/parachains/runtimes/collectives/collectives-westend/src/lib.rs:166]:
```suggestion
	pub const SS58Prefix: u8 = 42;
```

File: prdoc/pr_4857.prdoc [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      Introduces a new runtime API to help with conversions of XCM `Location` to the runtime's `AccountId`,
+      showing an Ss58 formatted address for easier verification.
+
+      Besides that, the `xcm-fee-payment-runtime-api` module was merged into the new `xcm-runtime-api`.
+      If you are using the `xcm-fee-payment-runtime-api` dependency, you just need to change it to `xcm-runtime-api`
+      and update the imports from `use xcm_fee_payment_runtime_api::*` to `use xcm_runtime_api::*`.
+
+crates:
+  - name: xcm-runtime-api
+    bump: patch
```
Comment (bkchr) [inline prdoc/pr_4857.prdoc:18]:
```suggestion
    bump: none
```

It is a new crate

File: polkadot/xcm/xcm-runtime-api/src/conversions.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use xcm::VersionedLocation;
+use xcm_executor::traits::ConvertLocation;
+
+sp_api::decl_runtime_apis! {
+	/// API for useful conversions between XCM `Location` and `AccountId`.
+	pub trait LocationToAccountApi {
+		/// Converts `Location` to `AccountId`.
+		fn convert_location(location: VersionedLocation) -> Result<sp_std::vec::Vec<u8>, Error>;
```
Comment (bkchr) [inline polkadot/xcm/xcm-runtime-api/src/conversions.rs:28]:
```suggestion
	pub trait LocationToAccountApi<AccountId> {
		/// Converts `Location` to `AccountId`.
		fn convert_location(location: VersionedLocation) -> Result<AccountId, Error>;
```

Don't hate me :see_no_evil:  Sorry, I think I said `Vec<u8>` before? But by having a propoer type here, downstream will be able to decode the type easily using the metadata.

**Review by bkchr — APPROVED** [review]

File: polkadot/xcm/xcm-runtime-api/src/conversions.rs (hunk 5 above)
Comment (bkontur) [reply polkadot/xcm/xcm-runtime-api/src/conversions.rs]: :D :D fixed here: https://github.com/paritytech/polkadot-sdk/pull/4857/commits/e39e6e742e95c157e38515ab30a24f25f451c034

File: polkadot/xcm/xcm-runtime-api/Cargo.toml [hunk 6]
```diff
@@ -1,36 +1,32 @@
 [package]
-name = "xcm-fee-payment-runtime-api"
+name = "xcm-runtime-api"
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-runtime-api/Cargo.toml:2]: Very nit: what about `xcm-runtime-apis`?

File: polkadot/xcm/xcm-runtime-api/tests/conversions.rs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+fn location_to_account_helper_with_multi_signature_works() {
+	type Signature = MultiSignature;
+	type AccountIdForConversions = <<Signature as Verify>::Signer as IdentifyAccount>::AccountId;
+	// We alias only `Location::parent()`
+	pub type LocationToAccountIdForConversions =
+		(xcm_builder::ParentIsPreset<AccountIdForConversions>,);
+
+	// Test unknown conversion for `Here` location
+	assert_err!(
+			LocationToAccountHelper::<
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-runtime-api/tests/conversions.rs:68]: fmt

File: prdoc/pr_4857.prdoc [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+doc:
+  - audience: Runtime User
+    description: |
+      Introduces a new runtime API to help with conversions of XCM `Location` to the runtime's `AccountId`,
+      showing an Ss58 formatted address for easier verification.
+
+      Besides that, the `xcm-fee-payment-runtime-api` module was merged into the new `xcm-runtime-api`.
+      If you are using the `xcm-fee-payment-runtime-api` dependency, you just need to change it to `xcm-runtime-api`
+      and update the imports from `use xcm_fee_payment_runtime_api::*` to `use xcm_runtime_api::*`.
```
Comment (franciscoaguirre) [inline prdoc/pr_4857.prdoc:14]: This looks like it's more suitable for the `Runtime Dev` audience

**Review by franciscoaguirre — APPROVED** [review]
🚀

File: polkadot/xcm/xcm-runtime-api/Cargo.toml (hunk 6 above)
Comment (bkontur) [reply polkadot/xcm/xcm-runtime-api/Cargo.toml:2]: ok, I will do that with all other comments when https://github.com/paritytech/polkadot-sdk/actions/runs/9679467294/job/26705514558?pr=4857 is finished

Comment (bkontur) [reply polkadot/xcm/xcm-runtime-api/Cargo.toml]: done: https://github.com/paritytech/polkadot-sdk/pull/4857/commits/540433c884d9c66ae5abfdd9a5d843510a0d8233

**Pull request merged**